File: Sources/Store.swift
import Foundation
import Darwin

final class DocumentStore {
    let root:URL
    init(root:URL?=nil) {
        self.root=root ?? ProcessInfo.processInfo.environment["BRANCH_HOME"].map {URL(fileURLWithPath:$0)} ?? FileManager.default.homeDirectoryForCurrentUser.appendingPathComponent("Library/Application Support/枝图",isDirectory:true)
    }
    static func validID(_ id:String) -> Bool {id.range(of:"^[a-zA-Z0-9-]{1,80}$",options:.regularExpression) != nil}
    func path(_ id:String) throws -> URL {guard Self.validID(id) else{throw BranchError("无效文档 ID")};return root.appendingPathComponent(id+".branch")}
    func prepare() throws {try FileManager.default.createDirectory(at:root,withIntermediateDirectories:true)}
    static func decode(_ data:Data) throws -> MindDocument {
        guard let raw=try JSONSerialization.jsonObject(with:data) as? [String:Any] else {throw BranchError("工程格式无效")}
        let d=MindDocument(raw:raw);try validate(d);return d
    }
    static func validate(_ d:MindDocument,images:Bool=true) throws {
        guard [1.0,2.0].contains(number(d.raw["version"])),validID(d.id),d.raw["title"] is String else {throw BranchError("工程版本、ID 或名称无效")}
        guard !d.nodes.isEmpty,d.nodes.count<=10000,Set(d.nodes.map(\.id)).count==d.nodes.count,!d.roots.isEmpty,(number(d.raw["version"])==2 || d.roots.count==1) else {throw BranchError("节点数量、ID 或主节点无效")}
        let by=Dictionary(uniqueKeysWithValues:d.nodes.map {($0.id,$0)})
        for n in d.nodes {
            guard validID(n.id),n.raw["title"] is String else{throw BranchError("节点 ID 或标题无效")}
            for key in ["notes","tags","link","color","shape","fontFamily","lineStyle"] where n.raw[key] != nil {guard n.raw[key] is String else{throw BranchError(key+" 必须是文本")}}
            guard (8...96).contains(n.fontSize),(60...240).contains(n.imageHeight) else{throw BranchError("字号或图片尺寸超出范围")}
            if let p=n.position {guard p.x.isFinite,p.y.isFinite else{throw BranchError("节点位置无效")}}
            var seen=Set<String>(),cur=n
            while let parent=cur.parent {
                guard !seen.contains(cur.id),seen.count<128,let next=by[parent] else {throw BranchError("层级循环、过深或父节点不存在")}
                seen.insert(cur.id);cur=next
            }
            let image=n.text("image")
            if images && !image.isEmpty {guard image.hasPrefix("data:image/jpeg;base64,"),let data=Data(base64Encoded:String(image.dropFirst(23))),data.starts(with:[255,216,255]),data.suffix(2)==Data([255,217]) else{throw BranchError("工程图片必须是有效 JPG")}}
        }
        for c in d.connections {guard let from=c["from"] as? String,let to=c["to"] as? String,by[from] != nil,by[to] != nil else{throw BranchError("关系线引用不存在的节点")}}
    }
    func read(_ id:String) throws -> MindDocument {
        let url=try path(id)
        do {return try Self.decode(Data(contentsOf:url))} catch {
            let backup=url.deletingPathExtension().appendingPathExtension("backup")
            guard FileManager.default.fileExists(atPath:backup.path) else{throw error}
            var d=try Self.decode(Data(contentsOf:backup));d.raw["_recovered"]=true;return d
        }
    }
    struct Entry {let id:String;let title:String;let revision:Int;let modified:Date}
    func list() throws -> [Entry] {
        try prepare()
        return try FileManager.default.contentsOfDirectory(at:root,includingPropertiesForKeys:[.contentModificationDateKey]).filter {$0.pathExtension=="branch"}.map {url in
            let d=try? read(url.deletingPathExtension().lastPathComponent)
            return Entry(id:url.deletingPathExtension().lastPathComponent,title:d?.title ?? url.lastPathComponent+"（文件损坏）",revision:d?.revision ?? -1,modified:(try? url.resourceValues(forKeys:[.contentModificationDateKey]).contentModificationDate) ?? .distantPast)
        }.sorted {$0.modified>$1.modified}
    }
    func save(_ document:MindDocument,expected:Int) throws -> MindDocument {
        try Self.validate(document);try prepare()
        let fd=Darwin.open(root.appendingPathComponent(".lock").path,O_CREAT|O_RDWR,0o600)
        guard fd>=0 else{throw BranchError("无法打开文档锁")};defer{flock(fd,LOCK_UN);Darwin.close(fd)}
        guard flock(fd,LOCK_EX)==0 else{throw BranchError("无法锁定工程")}
        let url=try path(document.id),old=FileManager.default.fileExists(atPath:url.path) ? try read(document.id):nil
        guard (old?.revision ?? 0)==expected else{throw BranchError("版本冲突：工程已由 MCP 或其他窗口修改。请导出副本后重新载入。")}
        if let old {try atomic(old.data(),to:url.deletingPathExtension().appendingPathExtension("backup"))}
        var d=document;d.revision=expected+1;d.raw.removeValue(forKey:"_recovered");try atomic(d.data(),to:url);return d
    }
    private func atomic(_ data:Data,to url:URL) throws {
        let tmp=root.appendingPathComponent(".pending-"+UUID().uuidString)
        let fd=Darwin.open(tmp.path,O_CREAT|O_EXCL|O_WRONLY,0o600)
        guard fd>=0 else{throw BranchError("无法创建临时工程")}
        defer{Darwin.close(fd);try? FileManager.default.removeItem(at:tmp)}
        try data.withUnsafeBytes {bytes in
            var written=0
            while written<data.count {let count=Darwin.write(fd,bytes.baseAddress!.advanced(by:written),data.count-written);guard count>0 else{throw BranchError("工程写入失败")};written+=count}
        }
        guard fsync(fd)==0,rename(tmp.path,url.path)==0 else{throw BranchError("工程写入失败")}
        let dir=Darwin.open(root.path,O_RDONLY);if dir>=0{_ = fsync(dir);Darwin.close(dir)}
    }
}
final class StoreWatcher {
    private var source:DispatchSourceFileSystemObject?
    init(url:URL,onChange:@escaping ()->Void) throws {
        let fd=Darwin.open(url.path,O_EVTONLY);guard fd>=0 else{throw BranchError("无法监听工程目录")}
        let source=DispatchSource.makeFileSystemObjectSource(fileDescriptor:fd,eventMask:[.write,.rename,.delete],queue:.main)
        source.setEventHandler(handler:onChange);source.setCancelHandler{Darwin.close(fd)};source.resume();self.source=source
    }
    deinit{source?.cancel()}
}

File: scripts/store.py
#!/usr/bin/env python3
"""Local document store shared by the app, CLI and MCP. No network access."""
import os, sys, json, uuid, tempfile, fcntl, base64, re
from pathlib import Path
from contextlib import contextmanager
ROOT = Path(os.environ.get('BRANCH_HOME', str(Path.home() / 'Library/Application Support/枝图')))
def validate(d):
    if d.get('version') not in (1,2): raise ValueError('不支持的工程版本')
    if not re.fullmatch(r'[a-zA-Z0-9-]{1,80}', d.get('id','')): raise ValueError('无效文档 ID')
    if not isinstance(d.get('title'),str): raise ValueError('文档名称必须是文本')
    nodes=d.get('nodes',[])
    if not nodes or len(nodes)>10000: raise ValueError('节点数量必须在 1–10000 之间')
    ids={n['id'] for n in nodes}
    if len(ids)!=len(nodes): raise ValueError('节点 ID 重复')
    by={n['id']:n for n in nodes}
    roots=sum(n.get('parent') is None for n in nodes)
    if roots<1 or (d['version']==1 and roots!=1): raise ValueError('至少保留一个主节点；多主节点工程需要 version 2')
    for n in nodes:
        if not re.fullmatch(r'[a-zA-Z0-9-]{1,80}',n['id']): raise ValueError('无效节点 ID')
        for key in ('notes','tags','link','color','shape','fontFamily','lineStyle'):
            if key in n and not isinstance(n[key],str): raise ValueError(key+' 必须是文本')
        for key,lo,hi in [('fontSize',8,96),('imageSize',60,240)]:
            if key in n and not lo <= float(n[key]) <= hi: raise ValueError(key+' 超出允许范围')
        if not isinstance(n.get('title'),str): raise ValueError('节点标题必须是文本')
        seen=set(); cur=n
        while cur.get('parent') is not None:
            if cur['id'] in seen: raise ValueError('不能形成循环层级')
            seen.add(cur['id'])
            if len(seen)>128: raise ValueError('节点层级超过 128 层')
            if cur['parent'] not in by: raise ValueError('父节点不存在')
            cur=by[cur['parent']]
        if n.get('image'):
            s=n['image']
            if not s.startswith('data:image/jpeg;base64,'): raise ValueError('工程图片必须是 JPG')
            raw=base64.b64decode(s.split(',',1)[1],validate=True)
            if not raw.startswith(b'\xff\xd8\xff') or not raw.endswith(b'\xff\xd9'): raise ValueError('JPG 数据损坏')
    for c in d.get('connections',[]):
        if c.get('from') not in ids or c.get('to') not in ids: raise ValueError('关系线引用不存在的节点')
    return d

def path(docid):
    if not re.fullmatch(r'[a-zA-Z0-9-]{1,80}',docid): raise ValueError('无效文档 ID')
    return ROOT/(docid+'.branch')
@contextmanager
def lock():
    ROOT.mkdir(parents=True,exist_ok=True)
    with open(ROOT/'.lock','a') as f:
        fcntl.flock(f,fcntl.LOCK_EX)
        yield

def atomic(p,data):
    fd,tmp=tempfile.mkstemp(dir=p.parent,prefix='.pending-')
    try:
        with os.fdopen(fd,'wb') as f: f.write(data); f.flush(); os.fsync(f.fileno())
        os.replace(tmp,p)
        fd=os.open(p.parent,os.O_RDONLY)
        try: os.fsync(fd)
        finally: os.close(fd)
    finally:
        if os.path.exists(tmp): os.unlink(tmp)

def read(docid):
    p=path(docid)
    try: return validate(json.loads(p.read_text()))
    except Exception as original:
        backup=p.with_suffix('.backup')
        if not backup.exists(): raise original
        d=validate(json.loads(backup.read_text()))
        d['_recovered']=True
        return d

def save(d,expected):
    validate(d); p=path(d['id'])
    with lock():
        old=read(d['id']) if p.exists() else None
        current=old.get('revision',0) if old else 0
        if current!=expected: raise ValueError('版本冲突：工程已由其他窗口或 MCP 修改。请先另存本地副本，再重新载入。')
        if old: atomic(p.with_suffix('.backup'),json.dumps(old,ensure_ascii=False).encode())
        d=dict(d); d.pop('_recovered',None); d['revision']=current+1
        atomic(p,json.dumps(d,ensure_ascii=False).encode())
    return d

def dispatch(a):
    op=a['op']
    if op=='list':
        ROOT.mkdir(parents=True,exist_ok=True); items=[]
        for p in ROOT.glob('*.branch'):
            try:
                d=read(p.stem); items.append({'id':d['id'],'title':d['title'],'revision':d['revision'],'modified':p.stat().st_mtime})
            except Exception: items.append({'id':p.stem,'title':p.stem+'（文件损坏）','revision':-1})
        return sorted(items,key=lambda x:x.get('modified',0),reverse=True)
    if op=='validate': return validate(a['document'])
    if op=='read': return read(a['id'])
    if op=='save': return save(a['document'],a.get('expected',0))
    if op=='import':
        d=validate(a['document']); d['id']=str(uuid.uuid4()); d['revision']=0
        return save(d,0)
    raise ValueError('未知操作')
if __name__=='__main__':
    try: print(json.dumps({'ok':True,'result':dispatch(json.load(sys.stdin))},ensure_ascii=False))
    except Exception as e: print(json.dumps({'ok':False,'error':str(e)},ensure_ascii=False))
